enterprise/server/registry/registry.go

```diff
@@ -737,10 +718,155 @@ type registry struct {
 	workDeduper singleflight.Group
 }
 
+// checkAccess whether the supplied credentials are sufficient to retrieve
+// the provided img.
+func (r *registry) checkAccess(ctx context.Context, imgRef ctrname.Reference, img v1.Image, authenticator authn.Authenticator) error {
+	// Check if we have access to all the layers.
+	// Obtaining the size of a layer requires a HEAD request to the remote repo.
+	layers, err := img.Layers()
+	if err != nil {
+		return status.UnknownErrorf("could not get layers for image: %s", err)
+	}
+	eg, egCtx := errgroup.WithContext(ctx)
+	remoteOpts := []remote.Option{remote.WithContext(egCtx)}
+	if authenticator != nil {
+		remoteOpts = append(remoteOpts, remote.WithAuth(authenticator))
+	}
+	for _, layerInfo := range layers {
+		layerInfo := layerInfo
+		eg.Go(func() error {
+			d, err := layerInfo.Digest()
+			if err != nil {
+				return err
+			}
+			layerRef := imgRef.Context().Digest(d.String())
+			l, err := remote.Layer(layerRef, remoteOpts...)
+			if err != nil {
+				return err
+			}
+			_, err = l.Size()
+			return err
+		})
+	}
```

Comment:
i don't totally understand how many remote requests this is making -- but is it really necessary to fetch all the layers in order to check access?